### PR TITLE
Milestone game logic

### PR DESCRIPTION
**Pull request by mricanho**
In this pull request, we work on the logic aspect of the game. We added a win condition and some conditions to prevent the players to do invalid moves.

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🤓 

### Hello team 😃,

**Great job on your project. However, there are a few fixes needed before the project can be approved.**

1). **The linter checks are not working on your pull request.**
linters

2). **Fix the following behaviours.**

- The game should not allow a player to play a move that has already been played.
- Please fix the message for when a player enters a wrong move. It tells the player to choose a number between `0-9` instead of `1-9`.

- Display the status of the board after a win or draw as well.

3). **All the logic should be inside the `lib` folder. You created most of your logic in the `bin` folder.**

4). **Do not use one class for everything like the way you did with `InitGame`. Create at least three classes. You can have classes for the Board, Players and game logic(to check for valid and invalid inputs). Remember to create them in `lib` folder.**

### Happy coding 🥂 

**PS: You can reach me on slack (Rose Wanjohi).**

File: lib/logic.rb
Comment (SpaYco) [inline lib/logic.rb]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: lib/logic.rb
Comment (SpaYco) [inline lib/logic.rb]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: lib/logic.rb
Comment (SpaYco) [inline lib/logic.rb]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: lib/logic.rb
Comment (SpaYco) [inline lib/logic.rb]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: bin/main.rb
Comment (SpaYco) [inline bin/main.rb]: - [x] All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: bin/main.rb
Comment (SpaYco) [inline bin/main.rb]: - [x] All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: bin/main.rb
Comment (SpaYco) [inline bin/main.rb]: - [x] All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

- [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)) return something from lib, then `puts/print` it using main.

- [x] move your InitGame class to lib (make sure you remove all prints/puts or gets.chomp), since your InitGame is about the game, your WinCondition class should instead be a method inside it.

- [x] since you need to move the WinCondition class, you still only have one class so far, please add a Player or a Board class (or both) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] Remember, all logic should be in lib. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

File: lib/logic.rb
Comment (teekaytech) [inline lib/logic.rb]: There should be NO puts/print and gets.chomp outside main.rb file.

File: bin/main.rb
Comment (teekaytech) [inline bin/main.rb]: Use `require_relative`

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've done so well in fixing the changes required as stated by the previous reviewers.
However, some changes are required before this project can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] Avoid using puts/print and gets.chomp outside main.rb 
- You're having `puts` in your `logic.rb` script.
- [This article](https://infinum.com/the-capsized-eight/multiline-strings-ruby-2-3-0-the-squiggly-heredoc) might help in storing multiline strings in a variable (using heredoc), then return the variable.

- [ ] Although your game does not allow a player to move into an occupied position, you can do better by ensuring that the player is informed about the wrong move, and also ask the same player (who made the wrong move) to try again, instead of moving on with the next player.

- [ ] User `require_relative` to link classes/modules, and not `require`.



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
Hello @mricanho and @Wusinho 
Your program runs fine but there are still some changes that you need to implement
- [x] Firstly, please fix the linter error that you have.
- [x] The application does not exit/stop due to invalid input - it informs the player about it and asks for their move again. Currently, your program stops when the player enters an invalid move, please update accordingly.
- [x] In your main.rb file, you are implementing some methods that are logical and therefore shouldn't be in the main.rb. All logical methods should be in the lib directory.
- [x] You should have at least 2 classes for handling the logic of your game. Right now you have only one class for the game logic. You can add a board class and allow it to handle anything that has to do with manipulating the game board and/ or a player class that handles every logic related to the player.
- [x] You should try as much as possible to respect the single responsibility rule of methods. Your methods should do one and only one thing. For instance, the method `win` in your logic class should ideally just be responsible for only checking if there is a win or not but you have violated the SRR of methods by also checking for `draw` in that same method.

Please request another review once you fix this.

slack(@ebeagusamuel)

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
#### Hello team, 
your game currently works fine but there are still some changes you need to make as regards the implementation of the game.
- [x] In your main.rb file, you are implementing some methods that are logical and therefore shouldn't be in the main.rb. All logical methods should be in the lib directory.
-[ ] You should have at least 2 classes for handling the logic of your game. Right now you have only one class for the game logic. You can add a board class and allow it to handle anything that has to do with manipulating the game board and/ or a player class that handles every logic related to the player.
- [x] You should try as much as possible to respect the single responsibility rule of methods. Your methods should do one and only one thing. For instance, the method win in your WinCondition class should ideally just be responsible for only checking if there is a win or not but you have violated the SRR of methods by also checking for a draw in that same method.
- [x] - optionally, you can inform the players of the winner of the game, i.e print the name of the winner of the game.

Request another review when you are one making the changes

slack(@ebeagusamuel )

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work, team. 💻 👀 ✨ 

## However, there are some changes required 👀 🧰 🔨 🔧 

- There is a bug in your code. When I made an overlapping move on space 3, the game has allowed me to make two consecutive turns as player 'X'. This should not be possible. Please fix this before continuing:

image

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Amazing Progress! 👏  Both of you have done well on this project.

You are yet to fix the bug the last reviewer talked about. 

- [x] Check the method for an invalid move, if the player makes an invalid move, you should not call turn but rather, he/she should make another move.

I made all the moves just by making player 2 make all invalid moves.
Screen Shot 2020-12-12 at 6 16 46 AM
Also, I think, 'position has already been taken' is kind of appropriate, as long as moves are within 1-9 just that it's been played already.

- [x] Also, notice the DRAW in the screenshot? I think you are counting player moves instead of checking if all pieces on the board have been moved. Kindly update that too.

- [x] You're disabling a lot of Rubocop checks, it kind of defeats the purpose of learning good coding practice because Rubocop is there to ensure we write not just good code but readable by other developers. For example, your `#WinCondition` method could be further refactored into `horizontal_wins, vertical_wins, and diagonal_wins` and those long lines broken into different lines.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by SafaErden — APPROVED** [review]
## Status: Approved 🥳🎉🎊

**Hi TEAM, Nice job implementing the required changes! The game looks amazing. I know that takes a lot of time and patience.**
##

**Take into account the following before merging your PR🧐:**

-I am approving this project to prevent you from an infinite re-review loop. Please consider fixing the bug explained below before you merge this project.

If the input is out of boundaries(less than 0 or more than 9, it returns: 'The position has already been taken, please choose another number', but this is incorrect, If the given number is out of boundaries, you should inform the player about his/her invalid move.)

##

Questions or comments? Reach me on ([Twitter](https://twitter.com/safaerden)) or ([LinkedIn](https://www.linkedin.com/in/safaerden/)) 🗨️

Happy coding_ 👨‍💻 

**Safa ERDEN**

**Pull request merged**
